ethereum/contracts/IncentivizedReceiveChannel.sol

```diff
@@ -0,0 +1,173 @@
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "./Decoder.sol";
+
+contract IncentivizedReceiveChannel {
+    using Decoder for bytes;
+
+    uint256 public lastProcessedNonce;
+
+    uint256 public MAX_PAYLOAD_BYTE_SIZE = 1000;
+    uint256 public MAX_PAYLOAD_GAS_COST = 500000;
+    uint256 public EXTERNAL_CALL_COST = 21000;
+    uint256 public MAX_GAS_PER_MESSAGE =
+        EXTERNAL_CALL_COST + MAX_PAYLOAD_BYTE_SIZE + EXTERNAL_CALL_COST;
+
+    struct Message {
+        uint256 nonce;
+        string senderApplicationId;
+        address targetApplicationAddress;
+        bytes payload;
+    }
+
+    struct Commitment {
+        bytes commitmentHash;
+    }
+
+    struct CommitmentContents {
+        Message[] messages;
+    }
+
+    constructor() public {
+        lastProcessedNonce = 0;
+    }
+
+    event MessageDelivered(uint256 _nonce, bool _result);
+
+    function newParachainCommitment(
+        Commitment memory commitment,
+        CommitmentContents memory commitmentContents,
+        uint256 parachainBlockNumber,
+        bytes memory ourParachainMerkleProof,
+        bytes memory parachainHeadsMMRProof
+    ) public {
+        verifyCommitment(
+            commitment,
+            commitmentContents,
+            ourParachainMerkleProof,
+            parachainHeadsMMRProof
+        );
+        processCommitmentContents(commitmentContents);
+    }
+
+    function verifyCommitment(
+        Commitment memory commitment,
+        CommitmentContents memory commitmentContents,
+        bytes memory ourParachainMerkleProof,
+        bytes memory parachainHeadsMMRProof
+    ) internal returns (bool success) {
+        // Prove we can get the MMRLeaf that is claimed to contain our Parachain Block Header
+        // BEEFYLightClient.verifyMMRLeaf(parachainHeadsMMRProof)
+        // BeefyLightClient{
+        // verifyMMRLeaf(parachainHeadsMMRProof) {
+        //MMRVerification.verifyInclusionProof(latestMMRRoot, parachainHeadsMMRProof)
+        // }
+        //}
+        //}
+        // returns mmrLeaf;
+
+        // Prove we can get the claimed parachain block header from the MMRLeaf
+        // allParachainHeadsMerkleTreeRoot = mmrLeaf.parachain_heads;
+        // MerkeTree.verify(allParachainHeadsMerkleTreeRoot, ourParachainMerkleProof)
+        // returns parachainBlockHeader
+
+        // Prove that the commitment is in fact in the parachain block header
+        // require(parachainBlockHeader.commitment == commitment)
+
+        // Validate that the commitment matches the commitment contents
+        require(
+            validateCommitment(commitment, commitmentContents),
+            "invalid commitment"
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >=
+                commitmentContents.messages.length * MAX_GAS_PER_MESSAGE,
+            "insufficient gas for delivery of all messages"
+        );
```

Comment:
We also need to account for gas that's used by the rest of the current function. Probably need to benchmark to find out how much we need to add exactly.

Reply:
yeh, will need to do proper gas-related review and changes in next pr/milestone

Reply:
added in https://github.com/Snowfork/polkadot-ethereum/pull/250